https://jira.mongodb.org/browse/SERVER-94430 Upstream respond:

> we intentionally do not ingest the environment PATH as this makes build
> reproducibility and hermiticity much harder. Instead, you should set the
> absolute paths to the tools you want to use on the command line, instead of
> relying on the PATH.

Gentoo bug https://bugs.gentoo.org/829340

In Gentoo, we have LLVM slotted and we put clang in /usr/lib/llvm/18/bin (or
whatever), not in /usr/bin, and if upstream strip PATH and construct it
themselves, they surely won't contain this location.

So we add PATH backup for scons.
diff --git a/SConstruct b/SConstruct
index 3d831c9..f07feeb 100644
--- a/SConstruct
+++ b/SConstruct
@@ -1201,7 +1201,7 @@ envDict = dict(BUILD_ROOT=buildDir,
 if get_option('build-tools') == 'next':
     SCons.Tool.DefaultToolpath.insert(0, os.path.abspath('site_scons/site_tools/next'))
 
-env = Environment(variables=env_vars, **envDict)
+env = Environment(variables=env_vars, ENV={'PATH': os.environ['PATH']}, **envDict)
 del envDict
 
 if get_option('cache-signature-mode') == 'validate':
--- a/SConstruct
+++ b/SConstruct
@@ -455,6 +455,7 @@ add_option(
 
 # Most of the "use-system-*" options follow a simple form.
 for pack in [
+    ('abseil-cpp',),
     (
         'asio',
         'ASIO',
@@ -1607,7 +1608,7 @@ def printLocalInfo():
 
 printLocalInfo()
 
-boostLibs = ["filesystem", "program_options", "system", "iostreams", "thread", "log"]
+boostLibs = ["filesystem", "program_options", "iostreams", "thread", "log"]
 
 onlyServer = len(COMMAND_LINE_TARGETS) == 0 or (len(COMMAND_LINE_TARGETS) == 1 and str(
     COMMAND_LINE_TARGETS[0]) in ["mongod", "mongos", "test"])
--- a/src/third_party/SConscript
+++ b/src/third_party/SConscript
@@ -21,7 +21,6 @@ tomcryptSuffix = '-1.18.2'
 variantSuffix = '-1.4.0'
 
 thirdPartyEnvironmentModifications = {
-    'abseil-cpp': {'CPPPATH': ['#/src/third_party/abseil-cpp/dist'], },
     'cares': {
         'CPPPATH': [
             '#src/third_party/cares/dist/include',
@@ -118,6 +117,11 @@ if not use_system_version_of_library('boost'):
         'CCFLAGS': ['$NOERROR_FOR_UNKNOWN_WARNING_OPTION_GEN'] if env.ToolchainIs('clang') else []
     }
 
+if not use_system_version_of_library('abseil-cpp'):
+    thirdPartyEnvironmentModifications['abseil-cpp'] = {
+        'CPPPATH' : ['#/src/third_party/abseil-cpp/dist'],
+    }
+
 if not use_system_version_of_library('snappy'):
     thirdPartyEnvironmentModifications['snappy'] = {
         'CPPPATH': ['#/src/third_party/snappy' + snappySuffix],
@@ -427,7 +431,6 @@ if use_system_version_of_library("boost"):
         boostEnv = boostEnv.Clone(SYSLIBDEPS=[
             env['LIBDEPS_BOOST_PROGRAM_OPTIONS_SYSLIBDEP'],
             env['LIBDEPS_BOOST_FILESYSTEM_SYSLIBDEP'],
-            env['LIBDEPS_BOOST_SYSTEM_SYSLIBDEP'],
             env['LIBDEPS_BOOST_IOSTREAMS_SYSLIBDEP'],
             env['LIBDEPS_BOOST_THREAD_SYSLIBDEP'],
             env['LIBDEPS_BOOST_LOG_SYSLIBDEP'],
@@ -442,22 +445,32 @@ else:
         boostDirectory + '/boost_iostreams',
         boostDirectory + '/boost_log',
         boostDirectory + '/boost_program_options',
-        boostDirectory + '/boost_system',
     ])
 
 boostEnv.ShimLibrary(name="boost")
 
-abseilDirectory = 'abseil-cpp'
 abseilEnv = env.Clone()
-abseilEnv.InjectThirdParty(libraries=['abseil-cpp'])
-abseilEnv.SConscript(abseilDirectory + '/SConscript', exports={'env': abseilEnv})
-abseilEnv = abseilEnv.Clone(LIBDEPS_INTERFACE=[
-    abseilDirectory + '/absl_city',
-    abseilDirectory + '/absl_hash',
-    abseilDirectory + '/absl_int128',
-    abseilDirectory + '/absl_low_level_hash',
-    abseilDirectory + '/absl_raw_hash_set',
-])
+if use_system_version_of_library("abseil-cpp"):
+    abseilEnv = abseilEnv.Clone(
+        SYSLIBDEPS=[
+            env['LIBDEPS_ABSL_CITY_SYSLIBDEP'],
+            env['LIBDEPS_ABSL_HASH_SYSLIBDEP'],
+            env['LIBDEPS_ABSL_INT128_SYSLIBDEP'],
+            env['LIBDEPS_ABSL_LOW_LEVEL_HASH_SYSLIBDEP'],
+            env['LIBDEPS_ABSL_RAW_HASH_SET_SYSLIBDEP'],
+        ])
+else:
+    abseilDirectory = 'abseil-cpp'
+    abseilEnv = abseilEnv.Clone()
+    abseilEnv.InjectThirdParty(libraries=['abseil-cpp'])
+    abseilEnv.SConscript(abseilDirectory + '/SConscript', exports={'env' : abseilEnv})
+    abseilEnv = abseilEnv.Clone(LIBDEPS_INTERFACE=[
+            abseilDirectory + '/absl_city',
+            abseilDirectory + '/absl_hash',
+            abseilDirectory + '/absl_int128',
+            abseilDirectory + '/absl_low_level_hash',
+            abseilDirectory + '/absl_raw_hash_set',
+        ])
 
 abseilEnv.ShimLibrary(name="abseil")
 
